ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -713,6 +646,152 @@ private AmbryReplica fetchBootstrapReplica(String partitionName) {
+    String partitionClass = replicaInfos.get(PARTITION_CLASS_STR);
+    AmbryPartition mappedPartition =
+        new AmbryPartition(Long.parseLong(partitionIdStr), partitionClass, helixClusterManagerQueryHelper);
+    AmbryPartition currentPartition =
+        partitionNameToAmbryPartition.putIfAbsent(mappedPartition.toPathString(), mappedPartition);
+    if (currentPartition == null) {
+      logger.info("Partition {} is currently not present in cluster map, a new partition is created", partitionIdStr);
+      currentPartition = mappedPartition;
+    }
+    // Check if data node or disk is in current cluster map, if not, set bootstrapReplica to null.
+    AmbryDataNode dataNode = instanceNameToAmbryDataNode.get(instanceName);
+    String mountPathAndDiskCapacityFromHelix = replicaInfos.get(instanceName);
+    String[] segments = mountPathAndDiskCapacityFromHelix.split(DISK_CAPACITY_DELIM_STR);
+    String mountPath = segments[0];
+    String diskCapacityStr = segments.length >= 2 ? segments[1] : null;
+    Set<AmbryDisk> disks = dataNode != null ? ambryDataNodeToAmbryDisks.get(dataNode) : null;
+    Optional<AmbryDisk> potentialDisk =
+        disks != null ? disks.stream().filter(d -> d.getMountPath().equals(mountPath)).findAny() : Optional.empty();
+    if (potentialDisk.isPresent()) {
+      try {
+        AmbryDisk targetDisk = potentialDisk.get();
+        if (diskCapacityStr != null) {
+          // update disk capacity if bootstrap replica info contains disk capacity in bytes.
+          targetDisk.setDiskCapacityInBytes(Long.parseLong(diskCapacityStr));
+        }
+        // A bootstrap replica is always ReplicaSealedStatus#NOT_SEALED.
+        bootstrapReplica = new AmbryServerReplica(clusterMapConfig, currentPartition, targetDisk, true, replicaCapacity,
+            ReplicaSealStatus.NOT_SEALED);
+      } catch (Exception e) {
+        logger.error("Failed to create bootstrap replica for partition {} on {} due to exception: ", partitionIdStr,
+            instanceName, e);
+        bootstrapReplica = null;
+      }
+    } else {
+      logger.error(
+          "Either datanode or disk that associated with bootstrap replica is not found in cluster map. Cannot create the replica.");
+    }
+    return bootstrapReplica;
+  }
+
+  /**
+   * Gets bootstrap replica in Helix full auto mode.
+   * @param partitionIdStr the partition id string
+   * @param dataNodeId the {@link DataNodeId} on which bootstrap replica is placed
+   * @return {@link ReplicaId} if there is a new replica satisfying given partition and data node. {@code null} otherwise.
+   */
+  private ReplicaId getBootstrapReplicaInFullAuto(String partitionIdStr, DataNodeId dataNodeId) {
+    try {
+      AmbryDisk disk = getDiskForBootstrapReplica((AmbryDataNode) dataNodeId);
+      if (disk == null) {
+        logger.error("No Disk is available to host bootstrap replica. Cannot create the replica.");
+        return null;
+      }
+      AmbryPartition mappedPartition =
```

Comment:
nit: let's log a trace log here so we know which disk is picked for this bootstrap replica.

Reply:
Done